File: src/auto_tagger/llm/prompts.py
"""Prompt builders for LLM tagging decisions."""

from __future__ import annotations

import json
from typing import Any

from auto_tagger.core.metadata import TrackMetadata
from auto_tagger.integrations.candidates import AlbumCandidate, LookupRequest

def build_selection_messages(
    request: LookupRequest,
    candidates: list[AlbumCandidate],
    max_candidates: int = 5,
) -> list[dict[str, str]]:
    """Build compact messages for candidate selection."""
    payload = {
        "artist_hint": request.artist_hint,
        "album_hint": request.album_hint,
        "track_count": len(request.tracks),
        "track_titles": [track.title for track in request.tracks if track.title],
        "candidates": [
            _candidate_summary(index, candidate)
            for index, candidate in enumerate(candidates[:max_candidates])
        ],
    }
    return [
        {
            "role": "system",
            "content": (
                "Select the best album candidate for audio tagging. "
                "Return only JSON with selected_index, confidence, and reason. "
                "Use selected_index null when all candidates are poor."
            ),
        },
        {"role": "user", "content": json.dumps(payload, ensure_ascii=False, sort_keys=True)},
    ]


def build_fallback_messages(
    request: LookupRequest,
    folder_candidate: AlbumCandidate,
    current_metadata: list[TrackMetadata],
) -> list[dict[str, str]]:
    """Build compact messages for fallback tag generation."""
    payload: dict[str, Any] = {
        "artist_hint": request.artist_hint,
        "album_hint": request.album_hint,
        "folder_candidate": _candidate_summary(0, folder_candidate),
        "current_tracks": [
            {
                "title": metadata.title,
                "artist": metadata.artist,
                "album": metadata.album,
                "track_number": metadata.track_number,
            }
            for metadata in current_metadata
        ],
    }
    return [
        {
            "role": "system",
            "content": (
                "Generate conservative fallback music tags as JSON. "
                "Do not invent MusicBrainz IDs. Leave uncertain fields empty. "
                "Return artist, artists, album, album_artist, album_artists, "
                "tracks, genre, confidence, and reason. "
                "For genre, use Discogs-style comma-separated tags "
                "(e.g. 'Electronic, House, Deep House' or 'Rock, Alternative, Indie' "
                "or 'Stage & Screen, Score, Contemporary Classical'). "
                "Only include genre if you are confident in the classification; "
                "leave null otherwise."
            ),
        },
        {"role": "user", "content": json.dumps(payload, ensure_ascii=False, sort_keys=True)},
    ]


def _candidate_summary(index: int, candidate: AlbumCandidate) -> dict[str, Any]:
    return {
        "index": index,
        "source": candidate.source.value,
        "artist": candidate.artist,
        "album": candidate.album,
        "album_artist": candidate.album_artist,
        "year": candidate.year,
        "musicbrainz_albumid": candidate.musicbrainz_albumid,
        "distance": candidate.distance,
        "tracks": [
            {
                "title": track.title,
                "track_number": track.track_number,
                "disc_number": track.disc_number,
            }
            for track in candidate.tracks[:20]
        ],
    }


def build_folder_extraction_messages(
    folder_name: str,
    parent_name: str | None = None,
) -> list[dict[str, str]]:
    """Build messages for extracting structured metadata from a folder name.

    The deterministic parser couldn't extract clean artist/album/year from
    this folder name (e.g., dot-separated ``Year.Artist.Album`` convention
    or multi-artist names). The LLM is asked to extract the fields.

    The ``parent_name`` (if provided) gives context — for CD subdirectories
    like ``Album CD1``, the parent folder ``Year.Artist.Album 3CD`` contains
    the year and artist info.
    """
    payload: dict[str, Any] = {
        "folder_name": folder_name,
        "instruction": (
            "Extract the artist, album name, and release year from this music folder name. "
            "If the folder name also indicates a disc number (e.g., CD1, Disc 2), "
            "include it. Return only JSON with these fields: "
            "artist (str), album (str), year (str or null), disc (str or null)."
        ),
    }
    if parent_name:
        payload["parent_name"] = parent_name

    return [
        {
            "role": "system",
            "content": (
                "You extract structured music metadata from folder names. "
                "Handle Chinese naming conventions like Year.Artist.Album "
                "and multi-artist folders (e.g., '2006.Artist1.Artist2.Album'). "
                "Return only valid JSON."
            ),
        },
        {"role": "user", "content": json.dumps(payload, ensure_ascii=False, sort_keys=True)},
    ]


# ── Audit prompts ─────────────────────────────────────────────────────────────


def build_audit_messages(
    album_artist_hint: str | None,
    album_hint: str | None,
    tracks: list[TrackMetadata],
    filenames: list[str],
) -> list[dict[str, str]]:
    """Build messages for LLM metadata audit.

    Audits per track: artist, title, album, album_artist, artists, path (filename).
    Includes positive and negative examples in the system prompt.
    """
    track_data = []
    for i, (meta, fn) in enumerate(zip(tracks, filenames)):
        track_data.append({
            "index": i,
            "path": fn,
            "title": meta.title or "",
            "artist": meta.artist or "",
            "album": meta.album or "",
            "album_artist": meta.album_artist or "",
            "artists": ", ".join(meta.artists) if meta.artists else "",
        })

    payload = {
        "album_folder": album_hint or "",
        "artist_folder": album_artist_hint or "",
        "tracks": track_data,
    }

    return [
        {
            "role": "system",
            "content": (
                "You audit music track metadata. The **primary principle** is that "
                "the file path must match the metadata: the album folder name should "
                "match the `album` field, the parent artist folder should match the "
                "`album_artist` field, and each filename should match its `title` field.\n\n"
                "The input provides:\n"
                "- `album_folder`: the album directory name (authoritative source for album name)\n"
                "- `artist_folder`: the parent directory name (authoritative source for album_artist)\n"
                "- `tracks[].path`: the filename (authoritative source for track title)\n"
                "- `tracks[].album`, `tracks[].album_artist`, `tracks[].title`: current metadata values\n\n"
                "Rules:\n"
                "1. 'error' means a field is clearly wrong or missing.\n"
                "2. 'warning' means the field might be wrong (typo, inconsistent capitalization, "
                "mismatched artist/album_artist convention).\n"
                "3. 'correct' means the field looks right.\n"
                "4. **Compare `tracks[].album` with `album_folder`.** If the album folder suggests "
                "a different album name than what is tagged, flag the `album` field. "
                "The folder name is authoritative.\n"
                "5. **Compare `tracks[].album_artist` with `artist_folder`.** If the parent "
                "artist directory suggests a different artist than what is tagged, "
                "flag the `album_artist` field. The folder name is authoritative.\n"
                "6. **Compare `tracks[].title` with `tracks[].path` (filename).** "
                "If the filename suggests a different title (after stripping track numbers, "
                "separators, and extensions), flag the `title` field.\n"
                "7. If artist != album_artist and artists is empty, flag artists as warning.\n"
                "8. Don't flag empty album_artist on single-artist albums.\n"
                "9. Be conservative — only flag when you have reasonable confidence.\n"
                "10. Title casing variations ('Come Together' vs 'come together') are warnings, "
                "not errors.\n"
                "11. For Chinese tracks: judge the correct character script "
                "(Simplified vs Traditional) based on the filename. The filename "
                "is the authoritative source for which script to use. If the "
                "filename uses Simplified Chinese, suggest Simplified Chinese "
                "in the title field; if it uses Traditional, suggest Traditional.\n"
                "12. **For every track with a warning or error, provide the complete "
                "corrected metadata in the `corrected` field.** Populate `corrected` "
                "with all the metadata fields that the track SHOULD have — title, "
                "artist, artists, album, album_artist, year, genre. Only include "
                "fields that are relevant (you don't need to repeat values that "
                "are already correct; the code will merge your corrected values "
                "with the existing metadata).\n"
                "13. The `suggestion` field is still used for per-field display "
                "but `corrected` is what gets written to the file.\n\n"
                "Examples of path-metadata matches (no issues):\n"
                "- album_folder='OK Computer', album='OK Computer' → correct\n"
                "- artist_folder='Radiohead', album_artist='Radiohead' → correct\n"
                "- path='01. Karma Police.flac', title='Karma Police' → correct\n"
                "\n"
                "Examples of path-metadata mismatches (with corrected metadata):\n"
                "- album_folder='OK Computer', album='OK Computer 1997' (wrong) → "
                "{ \"index\": 0, \"field\": \"album\", \"status\": \"error\", "
                "\"message\": \"Album tag 'OK Computer 1997' does not match folder name 'OK Computer'\", "
                "\"suggestion\": \"OK Computer\", "
                "\"corrected\": { \"album\": \"OK Computer\" } }\n"
                "- artist_folder='Pink Floyd', album_artist='Pink Floyd' but artist='David Gilmour' → "
                "{ \"index\": 0, \"field\": \"artist\", \"status\": \"warning\", "
                "\"message\": \"Track artist 'David Gilmour' differs from album_artist 'Pink Floyd', "
                "which matches the artist folder\", "
                "\"suggestion\": \"Pink Floyd\", "
                "\"corrected\": { \"artist\": \"Pink Floyd\" } }\n"
                "- path='01. 我爱的人.flac', title='I Love You' (English, not matching filename) → "
                "{ \"index\": 0, \"field\": \"title\", \"status\": \"error\", "
                "\"message\": \"Title 'I Love You' does not match filename '01. 我爱的人.flac'\", "
                "\"suggestion\": \"我爱的人\", "
                "\"corrected\": { \"title\": \"我爱的人\" } }\n"
                "- filename='03 - Bohemian Rhapsody.flac', title='Bohemian Rhapsody (Remastered 2011)' → "
                "{ \"index\": 0, \"field\": \"title\", \"status\": \"warning\", "
                "\"message\": \"Title 'Bohemian Rhapsody (Remastered 2011)' may have extra suffix not in "
                "filename\", "
                "\"suggestion\": \"Bohemian Rhapsody\", "
                "\"corrected\": { \"title\": \"Bohemian Rhapsody\" } }\n\n"
                "Other common patterns (preserved from existing logic):\n"
                "- artist='Beatles' (missing 'The') → "
                "{ \"index\": 0, \"field\": \"artist\", \"status\": \"warning\", "
                "\"message\": \"Artist may be missing 'The'\", "
                "\"suggestion\": \"The Beatles\" }\n"
                "- year='20' (truncated) → "
                "{ \"index\": 0, \"field\": \"year\", \"status\": \"error\", "
                "\"message\": \"Year truncated to 2 digits\", "
                "\"suggestion\": \"2020\" }\n"
                "- title is placeholder string → "
                "{ \"index\": 0, \"field\": \"title\", \"status\": \"error\", "
                "\"message\": \"Title is placeholder text, not real track name\" }\n\n"
                "Return only valid JSON. No markdown, no code fences, no extra text."
            ),
        },
        {"role": "user", "content": json.dumps(payload, ensure_ascii=False, sort_keys=True)},
    ]

File: frontend/electron/handlers/prompts.ts
/**
 * Prompt builders for LLM tagging decisions.
 * Ported from Python auto_tagger.llm.prompts.
 */

import type {
  AlbumCandidate,
  LookupRequest,
  TrackCandidate,
} from "./candidates";

// ── Selection ───────────────────────────────────────────────────────

export function buildSelectionMessages(
  request: LookupRequest,
  candidates: AlbumCandidate[],
  maxCandidates = 5,
): Array<{ role: string; content: string }> {
  const payload = {
    artist_hint: request.artistHint,
    album_hint: request.albumHint,
    track_count: request.tracks.length,
    track_titles: request.tracks.map((t) => t.title).filter(Boolean),
    candidates: candidates.slice(0, maxCandidates).map((c, i) => candidateSummary(i, c)),
  };

  return [
    {
      role: "system",
      content:
        "Select the best album candidate for audio tagging. " +
        "Return only JSON with selectedIndex, confidence, and reason. " +
        "Use selectedIndex null when all candidates are poor.",
    },
    {
      role: "user",
      content: JSON.stringify(payload, null, 2),
    },
  ];
}

// ── Fallback Tag Generation ─────────────────────────────────────────

export function buildFallbackMessages(
  request: LookupRequest,
  folderCandidate: AlbumCandidate,
  currentTracks: Array<{
    title?: string | null;
    artist?: string | null;
    album?: string | null;
    trackNumber?: number | null;
  }>,
): Array<{ role: string; content: string }> {
  const payload = {
    artist_hint: request.artistHint,
    album_hint: request.albumHint,
    folder_candidate: candidateSummary(0, folderCandidate),
    current_tracks: currentTracks,
  };

  return [
    {
      role: "system",
      content:
        "Generate conservative fallback music tags as JSON. " +
        "Do not invent MusicBrainz IDs. Leave uncertain fields empty. " +
        "Return artist, artists, album, albumArtist, albumArtists, " +
        "tracks, genre, confidence, and reason. " +
        "For genre, use Discogs-style comma-separated tags " +
        "(e.g. 'Electronic, House, Deep House' or 'Rock, Alternative, Indie' " +
        "or 'Stage & Screen, Score, Contemporary Classical'). " +
        "Only include genre if you are confident in the classification; " +
        "leave null otherwise.",
    },
    {
      role: "user",
      content: JSON.stringify(payload, null, 2),
    },
  ];
}

// ── Folder Extraction ───────────────────────────────────────────────

export function buildFolderExtractionMessages(
  folderName: string,
  parentName: string | null,
): Array<{ role: string; content: string }> {
  const payload: Record<string, string | null> = {
    folder_name: folderName,
  };
  if (parentName) {
    payload.parent_name = parentName;
  }

  return [
    {
      role: "system",
      content:
        "Extract music metadata from a folder name. " +
        "Return JSON with artist (the primary performing artist), " +
        "album (the release title), year (4-digit release year), " +
        "and disc (disc number if present). " +
        "Leave fields null if uncertain. " +
        "Do not invent information.",
    },
    {
      role: "user",
      content: JSON.stringify(payload, null, 2),
    },
  ];
}

// ── Tag Correction ──────────────────────────────────────────────────

/**
 * Build a prompt asking the LLM to correct/resolve album and track tags
 * by comparing folder names, existing file metadata, and the basic parser's hints.
 *
 * The LLM returns:
 *   - artist / albumArtist / album / year / genre — corrected album-level fields
 *   - tracks — per-track corrections (title, artist)
 *   - confidence — how confident the LLM is
 *
 * This output is used for:
 *   1. Corrected search params fed to MusicBrainz/Discogs API queries
 *   2. Fallback candidate when all API lookups return nothing
 */
/**
 * Extra context the caller may provide to help the LLM make better decisions.
 * All fields are optional — when absent the payload omits them.
 */
export interface TagCorrectionContext {
  /** Full directory path (gives the LLM structural context). */
  fullPath?: string;
  /** List of audio filenames in the directory. */
  filenames?: string[];
  /** Unique album tags found in the audio files. */
  existingAlbumTags?: string[];
  /** Unique artist tags found in the audio files. */
  existingArtistTags?: string[];
}

export function buildTagCorrectionMessages(
  folderName: string,
  parentName: string | null,
  parsedArtistHint: string | null,
  parsedAlbumHint: string | null,
  parsedYearHint: string | null,
  currentTracks: Array<{
    title?: string | null;
    artist?: string | null;
    album?: string | null;
    trackNumber?: number | null;
    genre?: string | null;
  }>,
  context?: TagCorrectionContext,
): Array<{ role: string; content: string }> {
  const payload: Record<string, unknown> = {
    folder_name: folderName,
    parsed_hints: {
      artist: parsedArtistHint,
      album: parsedAlbumHint,
      year: parsedYearHint,
    },
  };
  if (parentName) {
    payload.parent_name = parentName;
  }
  if (context?.fullPath) {
    payload.full_path = context.fullPath;
  }
  if (context?.filenames && context.filenames.length > 0) {
    payload.filenames = context.filenames;
  }
  if (context?.existingAlbumTags && context.existingAlbumTags.length > 0) {
    payload.existing_album_tags = context.existingAlbumTags;
  }
  if (context?.existingArtistTags && context.existingArtistTags.length > 0) {
    payload.existing_artist_tags = context.existingArtistTags;
  }
  if (currentTracks.length > 0) {
    payload.current_tracks = currentTracks.map((t, i) => ({
      index: i,
      title: t.title ?? null,
      artist: t.artist ?? null,
      album: t.album ?? null,
      track_number: t.trackNumber ?? null,
      genre: t.genre ?? null,
    }));
  }

  return [
    {
      role: "system",
      content:
        "Resolve correct music metadata by analyzing the folder name, " +
        "parent folder name, basic parser hints, and existing file tags.\n\n" +
        "CRITICAL: Your entire response must be ONLY a single JSON object. " +
        "Do NOT include any explanatory text, reasoning, analysis, " +
        "markdown formatting, or code fences before or after the JSON.\n\n" +
        "Rules:\n" +
        "1. Return JSON with artist, albumArtist, album, year, genre, tracks, and confidence.\n" +
        "2. The folder_name is the album directory (may include year prefix like '2009-', " +
        "format suffix like '[flac]', or other annotations). Strip these.\n" +
        "3. The parent_name (if present) is usually the artist or collection folder.\n" +
        "4. parsed_hints are from the basic folder-name parser — use them as a starting point.\n" +
        "5. current_tracks show existing metadata tags from the audio files — " +
        "use them to verify/correct artist, album, track titles.\n" +
        "6. If current_tracks have a consistent genre across all tracks, use it.\n" +
        "7. For genre, use Discogs-style comma-separated tags " +
        "(e.g. 'Hip Hop' or 'Electronic, House, Deep House').\n" +
        "8. Leave genre null if you are not confident.\n" +
        "9. For albumArtist: if all tracks share the same artist, use that. " +
        "If tracks have different artists (compilation/mixtape), use 'Various Artists'.\n" +
        "10. Per-track corrections: only include title/artist fields that differ " +
        "from the album defaults. Leave null for fields that match.\n" +
        "11. Do not invent MusicBrainz IDs. Leave uncertain fields null.\n" +
        "12. Set confidence (0.0-1.0) based on how sure you are about the correction.\n" +
        "13. parsed_hints may be WRONG — especially when the folder name contains a year " +
        "prefix followed by a number that looks like a month (e.g. '2009-100天' was " +
        "misparsed as '0天'). Always derive the album from folder_name by stripping " +
        "ONLY the 4-digit year+separator prefix. Never consume following album digits " +
        "as part of the date.\n" +
        "14. When folder_name uses 'Year - Artist - Album' or 'Year - Artist - Album " +
        "(format)' patterns, extract only the Album portion. Common separators: dash " +
        "(-), space, comma (,), underscore (_). Common format suffixes: (Lossless), " +
        "[FLAC], (24bit), (24bit-48Hz)(WAV).",
    },
    {
      role: "user",
      content: JSON.stringify(payload, null, 2),
    },
  ];
}

// ── Audit ──────────────────────────────────────────────────────────

export interface AuditContext {
  /** Discogs lookup alias (English search name for a Chinese artist). */
  discogsAlias?: string | null;
  reviewTargets?: Array<{
    index: number;
    field: string;
    current: string;
    expected?: string;
    evidence: string;
    reason: string;
  }>;
}

export function buildAuditMessages(
  albumArtistHint: string | null,
  albumHint: string | null,
  tracks: Array<{
    title?: string | null;
    artist?: string | null;
    album?: string | null;
    albumArtist?: string | null;
    albumArtists?: string[];
    artists?: string[];
    year?: string | null;
    genre?: string | null;
    trackNumber?: number | null;
    trackTotal?: number | null;
    discNumber?: number | null;
    discTotal?: number | null;
    path?: string;
  }>,
  filenames: string[],
  context?: AuditContext,
): Array<{ role: string; content: string }> {
  const trackData = tracks.map((meta, i) => ({
    index: i,
    path: filenames[i] ?? "",
    title: meta.title ?? "",
    artist: meta.artist ?? "",
    album: meta.album ?? "",
    album_artist: meta.albumArtist ?? "",
    album_artists: (meta.albumArtists ?? []).join(", "),
    artists: (meta.artists ?? []).join(", "),
    year: meta.year ?? "",
    genre: meta.genre ?? "",
    track_number: meta.trackNumber ?? null,
    track_total: meta.trackTotal ?? null,
    disc_number: meta.discNumber ?? null,
    disc_total: meta.discTotal ?? null,
  }));

  const payload: Record<string, unknown> = {
    album_folder: albumHint ?? "",
    artist_folder: albumArtistHint ?? "",
    preferred_artist_name: albumArtistHint ?? "",
    tracks: trackData,
    review_targets: context?.reviewTargets ?? [],
  };

  if (context?.discogsAlias) {
    payload.discogs_lookup_alias = context.discogsAlias;
  }

  return [
    {
      role: "system",
      content:
        "You audit music track metadata. Deterministic code has already checked " +
        "obvious path/tag mismatches, so you only review `review_targets`.\n\n" +
        "Primary principle: file path evidence must match metadata. The album " +
        "folder is evidence for album/year, the parent artist folder is evidence " +
        "for album artist, and each filename is evidence for title, artist, and " +
        "track/disc numbers.\n\n" +
        "Rules:\n" +
        "1. Review only fields listed in `review_targets`; do not invent findings for other fields.\n" +
        "2. Return one field at a time. Each result's `corrected` object may contain only that result's field.\n" +
        "3. Include `confidence` from 0.0 to 1.0 for every warning or error.\n" +
        "4. 'error' means the correction is clearly right; 'warning' means manual review is needed.\n" +
        "5. If evidence is ambiguous, return warning with no `corrected` value.\n" +
        "6. For genre, use Discogs-style comma-separated terms only when confident; otherwise warn.\n" +
        "7. **Chinese name preference**: When the `preferred_artist_name` " +
        "(from folder/tags) is in Chinese and a `discogs_lookup_alias` is " +
        "provided, prefer the Chinese name in corrected metadata. The English " +
        "alias is only a search aid, not a replacement tag value.\n" +
        "8. For Chinese tracks, judge Simplified vs Traditional script from the filename.\n\n" +
        "Output examples:\n" +
        '{ "index": 0, "field": "genre", "status": "warning", ' +
        '"message": "Genre is missing and cannot be inferred confidently from path evidence", ' +
        '"confidence": 0.4 }\n' +
        '{ "index": 1, "field": "title", "status": "error", ' +
        '"message": "Title has a clear typo compared with filename", ' +
        '"suggestion": "Karma Police", "confidence": 0.95, ' +
        '"corrected": { "title": "Karma Police" } }\n\n' +
        "Return only valid JSON. No markdown, no code fences, no extra text.",
    },
    {
      role: "user",
      content: JSON.stringify(payload),
    },
  ];
}

// ── Genre Fill ───────────────────────────────────────────────────────

/**
 * Build a prompt asking the LLM to provide genre for an album that already
 * has all other metadata resolved (artist, album, tracks from MusicBrainz/Discogs).
 * Only called when no source (Discogs, LLM enhance) provided genre.
 *
 * Returns JSON with just: { genre: string | null, confidence: number }
 */
export function buildGenreFillMessages(
  artist: string | null,
  album: string | null,
  trackTitles: string[],
): Array<{ role: string; content: string }> {
  const payload = {
    artist,
    album,
    track_titles: trackTitles,
  };

  return [
    {
      role: "system",
      content:
        "Classify the musical genre of this album. " +
        "Return only JSON with genre (string or null) and confidence (0.0-1.0). " +
        "Use Discogs-style comma-separated tags " +
        "(e.g. 'Electronic, House, Deep House' or 'Rock, Alternative, Indie' " +
        "or 'Stage & Screen, Score, Contemporary Classical'). " +
        "Leave genre null if you are not confident. " +
        "Do not invent information or guess from artist name alone.",
    },
    {
      role: "user",
      content: JSON.stringify(payload, null, 2),
    },
  ];
}

function candidateSummary(
  index: number,
  candidate: AlbumCandidate,
): Record<string, unknown> {
  return {
    index,
    artist: candidate.artist,
    album: candidate.album,
    year: candidate.year,
    genre: candidate.genre,
    track_count: candidate.tracks.length,
    track_titles: candidate.tracks.map((t) => t.title).filter(Boolean),
    source: candidate.source,
    musicbrainz_albumid: candidate.musicbrainzAlbumId,
    musicbrainz_artistid: candidate.musicbrainzArtistId,
  };
}
